drivers/audio/audio_dma.c

```diff
@@ -0,0 +1,607 @@
+                                        AUDIO_SAMP_RATE_192K;
+              break;
+          }
+
+        break;
+   }
+
+  /* Return the length of the audio_caps_s struct for validation of
+   * proper Audio device type.
+   */
+
+  return caps->ac_len;
+}
+
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+static int audio_dma_configure(struct audio_lowerhalf_s *dev,
+                               void *session,
+                               const struct audio_caps_s *caps)
+#else
+static int audio_dma_configure(struct audio_lowerhalf_s *dev,
+                               const struct audio_caps_s *caps)
+#endif
+{
+  struct audio_dma_s *audio_dma = (struct audio_dma_s *)dev;
+  struct dma_config_s cfg;
+  int ret = -EINVAL;
+
+  DEBUGASSERT(audio_dma && caps);
+  audinfo("ac_type: %d\n", caps->ac_type);
+
+  /* Process the configure operation */
+
+  switch (caps->ac_type)
+    {
+      case AUDIO_TYPE_OUTPUT:
+        if (audio_dma->playback)
+          {
+            memset(&cfg, 0, sizeof(struct dma_config_s));
+            cfg.direction = DMA_MEM_TO_DEV;
+            if (audio_dma->fifo_width)
+              cfg.dst_width = audio_dma->fifo_width;
+            else
+              cfg.dst_width = caps->ac_controls.b[2] / 8;
+            ret = DMA_CONFIG(audio_dma->chan, &cfg);
+          }
+        break;
+      case AUDIO_TYPE_INPUT:
+        if (!audio_dma->playback)
+          {
+            memset(&cfg, 0, sizeof(struct dma_config_s));
+            cfg.direction = DMA_DEV_TO_MEM;
+            if (audio_dma->fifo_width)
+              cfg.src_width = audio_dma->fifo_width;
+            else
+              cfg.src_width = caps->ac_controls.b[2] / 8;
+            ret = DMA_CONFIG(audio_dma->chan, &cfg);
+          }
+        break;
+      default:
+        ret = -ENOTTY;
+        break;
+    }
+
+  return ret;
+}
+
+static int audio_dma_shutdown(struct audio_lowerhalf_s *dev)
+{
+  struct audio_dma_s *audio_dma = (struct audio_dma_s *)dev;
+
+  /* apps enqueued buffers, but doesn't start. stop here to
+   * clear audio_dma->pendq.
+   */
+
+  audio_dma_stop(dev);
+
+  return OK;
+}
+
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+static int audio_dma_start(struct audio_lowerhalf_s *dev, void *session)
+#else
+static int audio_dma_start(struct audio_lowerhalf_s *dev)
+#endif
+{
+  struct audio_dma_s *audio_dma = (struct audio_dma_s *)dev;
+
+  return DMA_START_CYCLIC(audio_dma->chan, audio_dma_callback, audio_dma,
+                          audio_dma->dst_addr, audio_dma->src_addr,
+                          audio_dma->buffer_num * audio_dma->buffer_size,
+                          audio_dma->buffer_size);
+}
+
+#ifndef CONFIG_AUDIO_EXCLUDE_STOP
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+static int audio_dma_stop(struct audio_lowerhalf_s *dev, void *session)
+#else
+static int audio_dma_stop(struct audio_lowerhalf_s *dev)
+#endif
+{
+  struct audio_dma_s *audio_dma = (struct audio_dma_s *)dev;
+  struct ap_buffer_s *apb;
+
+  DMA_STOP(audio_dma->chan);
```

Comment:
Where is this macro defined?

Reply:
It's defined in https://github.com/apache/nuttx/pull/8773.